source/common/config/delta_subscription_impl.h

```diff
@@ -210,10 +225,28 @@ class DeltaSubscriptionImpl
       init_fetch_timeout_timer_.reset();
     }
   }
-  // A map from resource name to per-resource version.
-  std::unordered_map<std::string, std::string> resources_;
-  // The keys of resources_. Only tracked separately because std::map does not provide an iterator
-  // into just its keys, e.g. for use in std::set_difference.
+
+  // Use these helpers to avoid forgetting to update both at once.
+  void insertOrUpdateResourceVersion(const std::string& key, const std::string& val) {
+    resource_versions_[key] = val;
+    resource_names_.insert(key);
+  }
+
+  void eraseResourceVersion(const std::string& key) {
+    resource_versions_.erase(key);
+    resource_names_.erase(key);
+  }
+
+  // A map from resource name to per-resource version. Entries having EmptyVersion (which is just
+  // the empty string) are those which we are interested in, but have not gotten any version of from
+  // the server. We need these placeholders to know what to include in a delta discovery request:
+  // imagine we have lost interest in a resource the server never sent us. We need to inform the
+  // server that we no longer care, and if we didn't have an entry in resource_versions_ we wouldn't
+  // know to. On the other hand, when resource_versions_ is used to populate
+  // initial_resource_versions in a stream reconnect, we must leave these out.
```

Comment:
This is a bit subtle now.. we probably need to add a proto constraint to prohibit the empty version string on the xDS proto, which is technically a breaking change. Can we internally just use `absl::optional<std::string>` and avoid having any wire-level implications?

Reply:
Yeah, I actually considered exactly that. However, although the logic is now giving some slightly special meaning to a "" version, it doesn't meaningfully affect a server that sends "" versions. The only way this new logic makes Envoy behave (visibly to the server) differently is to omit things from the initial_resource_versions map. Omitting things from that is never wrong, just inefficient.

If you're worried about later changes accidentally making the weirdness around "" less benign, or even just maintainability of what is currently in this PR - that is a reasonable concern. I think forbidding "" in the protocol would be the right way to address that (even if it takes some community coordination), though, rather than absl::optional, which would make the code even more subtle. Consider that with optional, you would have "", nullopt, and [not in the map] all having different meanings. I think even with careful commenting, that would be hard to comprehend, even for the author e.g. 3 months later!

Reply:
Good point, but on balance I think I'd find it sub-optimal to have behavior that is inconsistent at the wire level based on subtleties like how "" is interepreted at the Envoy node for its internal book-keeping.

If you reckon `absl::optional` is unclear, how about some dedicated type along those lines, whose empty/null condition indicates 'WaitingForServer`. I'm thinking something like:
```
private:
  bool is_pending;
  std::string resource_version;
```

with the relevant methods. We don't want to be making wire changes to simplify something as easy to fix in code as this IMHO.

Reply:
Thanks for the suggestion, that's a good idea. I added a helper class along those lines, and renamed the helper functions. It all feels very clear and clean to me now; hopefully it will to you also!